src/core/tools/hpc/hpc_logger.cpp

```diff
@@ -0,0 +1,399 @@
+			const int line,
+			dsn_log_level_t log_level,
+			const char* title,
+			const char *fmt,
+			va_list args
+			)
+		{
+			if (s_hpc_log_tls_info.magic != 0xdeadbeef)
+			{
+				s_hpc_log_tls_info.buffer = (char*)malloc(_per_thread_buffer_bytes);
+				s_hpc_log_tls_info.next_write_ptr = s_hpc_log_tls_info.buffer;
+				s_hpc_log_tls_info.last_hdr = nullptr;
+				memset(s_hpc_log_tls_info.buffer, '\0', _per_thread_buffer_bytes);
+
+				hpc_log_manager::instance().put(::dsn::utils::get_current_tid(), &s_hpc_log_tls_info);
+				s_hpc_log_tls_info.magic = 0xdeadbeef;
+			}
+
+			// get enough write space >= 1K
+			if (s_hpc_log_tls_info.next_write_ptr + 1024 > s_hpc_log_tls_info.buffer + _per_thread_buffer_bytes)
+			{
+				//critical section begin
+				_write_list_lock.lock();
+
+				buffer_push(s_hpc_log_tls_info.buffer, s_hpc_log_tls_info.last_hdr);
+
+				if (_write_list.size() > 0) _write_list_cond.notify_one();
+				_write_list_lock.unlock();
+				//critical section end
+
+				s_hpc_log_tls_info.buffer = (char*)malloc(_per_thread_buffer_bytes);
+				s_hpc_log_tls_info.next_write_ptr = s_hpc_log_tls_info.buffer;
+				s_hpc_log_tls_info.last_hdr = nullptr;
+				memset(s_hpc_log_tls_info.buffer, '\0', _per_thread_buffer_bytes);
+				
+
+			}
+			char* ptr = s_hpc_log_tls_info.next_write_ptr;
+			char* ptr0 = ptr; // remember it
+			size_t capacity = static_cast<size_t>(s_hpc_log_tls_info.buffer + _per_thread_buffer_bytes - ptr);
+
+			// print verbose log header    
+			uint64_t ts = 0;
+			int tid = ::dsn::utils::get_current_tid();
+			if (::dsn::tools::is_engine_ready())
+				ts = dsn_now_ns();
+			char str[24];
+			::dsn::utils::time_ms_to_string(ts / 1000000, str);
+			auto wn = sprintf(ptr, "%s (%llu %04x) ", str, static_cast<long long unsigned int>(ts), tid);
+			ptr += wn;
+			capacity -= wn;
+
+			task* t = task::get_current_task();
+			if (t)
+			{
+				if (nullptr != task::get_current_worker())
+				{
+					wn = sprintf(ptr, "%6s.%7s%u.%016llx: ",
+						task::get_current_node_name(),
+						task::get_current_worker()->pool_spec().name.c_str(),
+						task::get_current_worker()->index(),
+						static_cast<long long unsigned int>(t->id())
+						);
+				}
+				else
+				{
+					wn = sprintf(ptr, "%6s.%7s.%05d.%016llx: ",
+						task::get_current_node_name(),
+						"io-thrd",
+						tid,
+						static_cast<long long unsigned int>(t->id())
+						);
+				}
+			}
+			else
+			{
+				wn = sprintf(ptr, "%6s.%7s.%05d: ",
+					task::get_current_node_name(),
+					"io-thrd",
+					tid
+					);
+			}
+
+			ptr += wn;
+			capacity -= wn;
+
+			// print body
+			wn = std::vsnprintf(ptr, capacity, fmt, args);
+			ptr += wn;
+			capacity -= wn;
+
+			// set binary entry header on tail
+			hpc_log_hdr* hdr = (hpc_log_hdr*)ptr;
+			hdr->log_break = 0;
+			hdr->length = 0;
+			hdr->magic = 0xdeadbeef;
+			hdr->ts = ts;
+			hdr->length = static_cast<int>(ptr - ptr0);
+			hdr->prev = s_hpc_log_tls_info.last_hdr;
+			s_hpc_log_tls_info.last_hdr = hdr;
```

Comment:
for tail log, last_hdr is necessary as it is circular buffer and the head of the buffer block is not necessarily the entry of the first log. However, for hpc_log, we don't need this any more, and even the log_hdr data structure is not necessary any more.